### ISSUE TITLE
Scope of Warn Reach reporting

### ISSUE THREAD

**Issue by jfdm** [idris-lang/Idris-dev#2404]
The [idris-containers](https://github.com/jfdm/idris-containers) package is currently built with the flag `--warnreach`. The precise semantics of what reach warns about, I am not 100% sure about. The phrase 'reachable but inaccessible' is not intuitive with out further explanation.

Regardless, when iris-containers is built and installed with this flag on, the idris compiler offers no complaints. However, if I run the tests warn reach suddenly starts to complain. I thought the flag would have picked up the warnings, or does bringing functions into scope through use make the check more thorough. Incidentally, I noticed this behaviour when compiler packages that depend on iris-containers when hunting down a seg-fault that may or may not be related to the warning.

**Comment by david-christiansen** [comment]
ping @ziman

**Comment by ahmadsalim** [comment]
ping @ziman

**Comment by ziman** [comment]
Yeah, the terminology was not settled at the time of writing that error message and it's quite confusing.

"Reachable but inaccessible" means that it's "reachable", i.e. potentially influencing the result of `main`; "inaccessible" means it was marked as "to be erased", i.e. dotted (or free implicit).

The reason that libraries alone don't produce warnings is because usage of erased stuff is okay and expected in type-level functions — but without `main`, we don't know *which* functions are type-level. That's why you don't get the warnings until you actually compile a program.

There are several solutions (ideally, all of them):
* better docs
* better error messages
* `%run_time` or so for library functions you want to check even without `main`

**Comment by jfdm** [comment]
Yeah, having a quick look at [erasure.rst](http://idris.readthedocs.io/reference/erasure.html), it does need 'refreshing'.
